src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/ValuesSources/WeaklyCachedRecoverableValueSource.cs

```diff
@@ -128,7 +128,11 @@ private void ResetRecoverySource(Task saveTask, T instance)
             {
                 using (Gate.DisposableWait(CancellationToken.None))
                 {
-                    _recoverySource = new AsyncLazy<T>(RecoverAsync, Recover, cacheResult: false);
+                    // Only assume the instance is saved if the saveTask completed successfully. If the save did not
+                    // complete, we can still rely on a constant value source to provide the instance.
+                    _recoverySource = saveTask.Status == TaskStatus.RanToCompletion
+                        ? new AsyncLazy<T>(RecoverAsync, Recover, cacheResult: false)
+                        : new ConstantValueSource<T>(instance);
```

Comment:
this makes sense.  though i worry about waht is breaking teh saveTask.  do we have insight on that?  Can we report/log failures of that as well so we can track down any other necessary fixes?

Reply:
We already do log NFE for failure of `saveTask`. @ryzngard could look up telemetry for it.